manager/integration/tests/test_settings.py

```diff
@@ -788,3 +793,260 @@ def test_setting_concurrent_rebuild_limit():  # NOQA
+
+    volume2.attach(hostId=lht_host_id)
+    wait_for_volume_healthy(client, volume2_name)
+
+    # Step 3-1
+    # Step 3-2
+    concurrent_rebuild_when_engine_live_upgrade(client,
+                                                volume1_name,
+                                                volume2_name)
+
+
+def concurrent_rebuild_when_engine_live_upgrade(client, volume1_name, volume2_name): # NOQA
+
+    default_img = common.get_default_engine_image(client)
+    default_img_name = default_img.name
+    default_img = wait_for_engine_image_ref_count(client, default_img_name, 2)
+    cli_v = default_img.cliAPIVersion
+    cli_minv = default_img.cliAPIMinVersion
+    ctl_v = default_img.controllerAPIVersion
+    ctl_minv = default_img.controllerAPIMinVersion
+    data_v = default_img.dataFormatVersion
+    data_minv = default_img.dataFormatMinVersion
+    engine_upgrade_image = common.get_upgrade_test_image(cli_v, cli_minv,
+                                                         ctl_v, ctl_minv,
+                                                         data_v, data_minv)
+
+    new_img = client.create_engine_image(image=engine_upgrade_image)
+    new_img_name = new_img.name
+    ei_status_value = get_engine_image_status_value(client, new_img_name)
+    new_img = wait_for_engine_image_state(client,
+                                          new_img_name,
+                                          ei_status_value)
+    assert new_img.refCount == 0
+    assert new_img.noRefSince != ""
+
+    default_img = common.get_default_engine_image(client)
+    default_img_name = default_img.name
+
+    volume2 = client.by_id_volume(volume2_name)
+
+    original_engine_image = volume2.engineImage
+    assert original_engine_image != engine_upgrade_image
+
+    delete_replica_on_test_node(client, volume1_name)
+    wait_for_rebuild_start(client, volume1_name)
+
+    assert volume2.engineImage == original_engine_image
+    assert volume2.currentImage == original_engine_image
+    volume2 = client.by_id_volume(volume2_name)
+    engine = get_volume_engine(volume2)
+    assert engine.engineImage == original_engine_image
+    assert engine.currentImage == original_engine_image
+
+    volume2.engineUpgrade(image=engine_upgrade_image)
+
+    # In a 2 minutes retry loop:
+    # verify that we see the case, volume 2 finished upgrading while volume
+    # is still rebuilding
+    # volume 2 finished upgrading if it meet the condition:
+    # *it is healthy
+    # *currentImage of volume2 == engine_upgrade_image
+    expect_case = False
+    for i in range(RETRY_COUNTS):
+        volume1 = client.by_id_volume(volume1_name)
+        volume2 = client.by_id_volume(volume2_name)
+        try:
+            if volume1.rebuildStatus[0].state == "in_progress" and \
+                    volume2.currentImage == engine_upgrade_image and \
+                    volume2["robustness"] == "healthy" and \
+                    len(volume2.replicas) == 3:
```

Comment:
Is the existing data large enough so that the rebuilding can be slower than the live upgrade?

Reply:
In my environment looks OK, I will follow pipeline result after approved